nixos/modules/services/web-apps/porn-vault/default.nix

```
@@ -0,0 +1,111 @@
+{
+  config,
+  pkgs,
+  lib,
+  ...
+}:
+
+let
+  cfg = config.services.porn-vault;
+  configFormat = pkgs.formats.json { };
+  inherit (lib)
+    mkIf
+    mkEnableOption
+    mkPackageOption
+    mkOption
+    getExe
+    literalExpression
+    types
+    ;
+in
+{
+  options = {
+    services.porn-vault = {
+      enable = lib.mkEnableOption "Porn-Vault";
+
+      package = lib.mkPackageOption pkgs "porn-vault" { };
+
+      autoStart = lib.mkOption {
+        type = lib.types.bool;
+        default = false;
+        description = ''
+          Whether to start porn-vault automatically.
+        '';
+      };
+
+      port = lib.mkOption {
+        type = lib.types.port;
+        default = 3000;
+        description = ''
+          Which port Porn-Vault will use.
+        '';
+      };
+
+      openFirewall = lib.mkOption {
+        type = lib.types.bool;
+        default = false;
+        description = ''
+          Whether to open the Porn-Vault port in the firewall.
+        '';
+      };
+
+      config = {
+        enable = mkEnableOption "Porn-Vault configuration";
+        settings = mkOption {
+          type = configFormat.type;
+          description = ''
+            Configuration for Porn-Vault. The attributes are serialized to JSON in config.json.
+
+            See https://gitlab.com/porn-vault/porn-vault/-/blob/dev/config.example.json
+          '';
+          default = import ./default-config.nix;
```

```suggestion
          default = defaultConfig;
          apply = lib.recursiveUpdate defaultConfig;
```

We can make sure the keys required are always present this way, even if the user tries to use `lib.mkForce` for example

---

Wdym?

---

Ohhh that’s cool!